sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -46,6 +46,20 @@ trait CheckAnalysis extends PredicateHelper {
     }).length > 1
   }
 
+  private def checkLimitClause(limitExpr: Expression): Unit = {
+    if (!limitExpr.foldable) {
```

Comment:
I'm saying maybe we can declare Limit as `case class GlobalLimit(limit: Int, child: LogicalPlan)`, but it's not a small change, we could think about it later.